File: java/SR-RBD/src/ddb/msg/HelloMessage.java
package ddb.msg;

import java.io.IOException;
import java.util.LinkedList;
import java.util.List;

import ddb.communication.DataInputStream;
import ddb.communication.DataOutputStream;
import ddb.db.TableVersion;
import ddb.msg.Message;

public class HelloMessage extends Message 
{
	private List<TableVersion> tables = new LinkedList<TableVersion>();

	
	public HelloMessage()
	{
		// empty
	}
	
	public HelloMessage(List<TableVersion> tables)
	{
		setTables(tables);
	}
	
	private void setTables(List<TableVersion> tables) {
		this.tables = tables;
	}

	private List<TableVersion> getTables() {
		return tables;
	}
	
	@Override
	public void fromBinary(DataInputStream s) throws IOException {
		// read count
		int count = s.readInt();
		setTables(new LinkedList<TableVersion>());
		
		// read contents
		for(int i = 0; i < count; ++i)
			getTables().add(new TableVersion(s));
		
	}

	@Override
	public MessageType getType() {
		return MessageType.HELLO_MESSAGE;
	}

	@Override
	public void toBinary(DataOutputStream s) throws IOException {
		// write length
		s.writeInt(getTables().size());
		
		// write tables
		for(TableVersion tv : getTables())
			tv.toBinary(s);
	}
}

File: trunk/C#/RBD/RBD/Msg/HelloMessage.cs
// + TODO check

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections;

using RBD.DB;
using RBD.Communication;

namespace RBD.Msg
{
    public class HelloMessage : Message
    {
        public int ListeningPort { get; set; }
        public IList<TableVersion> Tables { get; set; }

        public HelloMessage()
        {
            // empty
        }

        public HelloMessage(IList<TableVersion> tables, int port)
        {
            Tables = tables;
            ListeningPort = port;
        }

        override public void FromBinary(DataInputStream s) //throws IOException
        {
            // read port
            ListeningPort = s.readInt();

            // read count
            int count = s.readInt();
            Tables = new List<TableVersion>();  // TODO check -- nie jestem pewny czy jest to odpowiednik LinkedList z Javy

            // read contents
            while (count-- > 0)
                Tables.Add(new TableVersion(s));
        }

        override public MessageType GetMessageType()
        {
            return MessageType.HELLO_MESSAGE;
        }

        override public void ToBinary(DataOutputStream s) //throws IOException
        {
            // write port
            s.Write((int)ListeningPort);

            // write length
            s.Write((int)Tables.Count);

            // write tables
            foreach (TableVersion tv in Tables)
                tv.ToBinary(s);
        }
    }
}
